### ISSUE TITLE
[slack-status ] Authentication Loop—Doesn't work anymore

### ISSUE THREAD

**Issue by xmadscientist** [raycast/extensions#15917]
### Extension

https://www.raycast.com/petr/slack-status

### Raycast Version

1.88.3

### macOS Version

14.7.1

### Description

I keep getting put into an authentication loop.

![image](https://github.com/user-attachments/assets/c326e19f-058f-46fd-a09e-e109632dee75)

![image](https://github.com/user-attachments/assets/fb40815b-5b35-4bd7-a3ea-8ecb34edd09a)

And cannot actually set statuses anymore.

### Steps To Reproduce

1. Run slack-status from Raycast
2. When the prompt for re-connecting your Slack account shows up, click on the `Confirm` button.
3. Sign into your Slack workspace.
4. After signing in and authenticating, click `Allow`
5. You will see a success message saying that you were successfully authenticated
6. Raycast will open, but you will not be authenticated.
7. You will be looped back to step #2

### Current Behaviour

N/A—outlined above.

### Expected Behaviour

Get authenticated and start using slack-status from Raycast.

**Comment by pernielsentikaer** [comment]
CC @henrik

**Comment by henrik** [comment]
Just getting on a plane but will look soon.

@xmadscientist When you go through the auth flow, does Slack list dnd among the permissions?

Basically the code tried to get DND status and if it fails due to missing permissions, it logs you out and when you log back in, that should be with the right permissions.

**Comment by henrik** [comment]
Thanks! I tried the upgrade flow several times when writing it (having old permissions, running the extension, being logged out, logging back in) and never saw this issue 🤔 

Would be interesting to hear if this is an isolated issue – is it working for others?

Does Raycast show any error message (toast bottom left) when it happens? If so, can you expand it and tell me what it says?

If you're willing to debug further, you could go to `https://CHANGE_THIS.slack.com/apps/manage`, find Raycast among your installed apps, and confirm that the "Permissions" tab includes "dnd:read".

You could also go its "App details" tab, then "Configuration", then try revoking your authorization from there, and going through the flow again.

If that doesn't help, does @pernielsentikaer have any suggestions on Raycast debugging practices? Could we make an update with a lot of sprinkled `console.log`-ing, and is there then a way for a user to get at those logs?

**Comment by xmadscientist** [comment]
![image](https://github.com/user-attachments/assets/93afbf3d-0940-4364-a32f-a4a46dc73e3b)

It does include `dnd:read`!

> You could also go its "App details" tab, then "Configuration", then try revoking your authorization from there, and going through the flow again.

Trying this right now.

**Comment by xmadscientist** [comment]
> Does Raycast show any error message (toast bottom left) when it happens? If so, can you expand it and tell me what it says?

Nope! It doesn't show any error messages. In fact, it flashes the "Slack successfully connected" Raycast panel for a few frames before looping back to asking for authorization again.

**Comment by henrik** [comment]
Thanks! This is the relevant code: https://github.com/raycast/extensions/blob/1712079c9b98ae20ae5f25feb03ed197dd9719b6/extensions/slack-status/src/slack.ts#L22-L50

The idea with the loop is to reattempt the DND lookup (whether you've paused notifications) after being logged out and re-authenticating, but it sounds like you're somehow authenticating, but then looking up the DND info fails with a "missing_scope" error, so it logs you out and on and on it goes.

Just to rule it out, I now tried with a local production build (`npm run build`) to throw some other error (not "missing scope") and it did show up as a toast + did not cause a login loop. Just an empty status.

Did revoking the permissions change anything?

**Comment by thnk2wn** [comment]
I have the same issue. I an authenticate correctly but when I come back it still thinks I don't have correct permissions. I can briefly see my previous / current custom statuses beneath but can't use them.

<img width="750" alt="image" src="https://github.com/user-attachments/assets/f2e5813d-33a6-4830-a2bd-93d10295ae4b" />

**Comment by xmadscientist** [comment]
> Thanks! This is the relevant code:
> 
> https://github.com/raycast/extensions/blob/1712079c9b98ae20ae5f25feb03ed197dd9719b6/extensions/slack-status/src/slack.ts#L22-L50
> 
> The idea with the loop is to reattempt the DND lookup (whether you've paused notifications) after being logged out and re-authenticating, but it sounds like you're somehow authenticating, but then looking up the DND info fails with a "missing_scope" error, so it logs you out and on and on it goes.
> 
> Just to rule it out, I now tried with a local production build (`npm run build`) to throw some other error (not "missing_scope") and it did show up as a toast + did not cause a login loop. Just an empty status.
> 
> Did revoking the permissions change anything?

Unfortunately, revoking the permissions didn't help. :(

**Comment by henrik** [comment]
@PitNikola @pernielsentikaer Can you confirm the extension works for you?

**Comment by henrik** [comment]
The permissions screen has the message "Below are the default permissions configured for Raycast by the developer. Permissions installed might be different for each workspace."

So one possible explanation might be that the workspace does not grant this permission, even though it's a default permission. I'm not sure how to test that theory, though. Can you see any indications that "dnd:read" is denied by the workspace?

This is what my App details -> Configuration page looks like, but I'm not sure if it actually shows what the workspace allows. Does yours look any different, @xmadscientist?

<img width="1021" alt="Screenshot 2024-12-19 at 09 44 56" src="https://github.com/user-attachments/assets/32f64f5c-d07f-4f25-895c-ac5c1d635da4" />

**Comment by thnk2wn** [comment]
@henrik If it helps any...

https://github.com/user-attachments/assets/ac2ecb9b-7732-4839-809a-381b19f0d2c7

**Comment by henrik** [comment]
Thanks so much for the additional detail!

---

Regarding the "token_revoked" error mentioned by @thnk2wn, based on the line number I think this might be the `slackAuth.client.removeTokens()` call (are you able to confirm via the stated file and line number on your end?).

If that's the case, then having the extension silently ignoring that error *might* help, though I can't say I fully understand what's going on.

It succeeds getting the profile details, but then fails to get the DND details with "token_revoked". Perhaps there is still one unrevoked app/authentication without DND permissions, that succeeds in getting the profile; and there's also another app/authentication *with* DND permissions, that Raycast tries to use, but fails to because it's revoked on the Slack end?

---

@xmadscientist Thank you. You cut off the name of the app in the second screenshot, but are you saying that there's an app in Slack that got DND permissions in December 2024 but that is *not* named "Slack Status"? And there's also the "Slack Status" app which does *not* have DND permissions, and that was authorised in June 2023?

---

@pernielsentikaer Do you have context on how Raycast handles OAuth that might help understand this situation? Each extension using Slack should get its own access token, right? Two extensions using Slack with different permissions should be able to live side by side?

**Comment by thnk2wn** [comment]
@henrik I can't remember whether the error I posted was at load or when trying to change status but it appears the latter

<img width="640" alt="image" src="https://github.com/user-attachments/assets/aefdfcf6-4b94-4382-9894-b62701c2e096" />

**Comment by henrik** [comment]
Thanks!

My thinking so far is that I want to hear more from Raycast (e.g. Per) about how this is *supposed* to work – if there could be some kind of interference between Slack integrations, and if so, whether that's considered a bug.

I also think I should add a text to the "Missing permissions" dialogue saying something to the effect of "If you keep seeing this message, please log out manually in the extension settings and uninstall all Raycast apps via Slack's admin UI".

I appreciate that this is a frustrating situation, but I'm still not clear on precisely what's going on, so I can't reproduce it myself to fix it more gracefully than that. But it does seem to me like there is some leakiness to (my assumptions about) the abstractions that extensions lean on.

It would be great if @thnk2wn and @xmadscientist could confirm whether doing those things fixes the issue. So specifically:

1. In the "Slack Status" extension settings in Raycast, click "Logout" if available. (I think this might fix the "token_revoked" error.)
2. In the Slack web admin interface, select "Uninstall" on all "Raycast" apps
3. Open "Slack Status" again and go through the auth flow.

**Comment by thnk2wn** [comment]
@henrik Thanks, that worked with the caveat that clicking the Logout button didn't actually clear the token, I had to wipe that out manually.

I'm good with this workaround but going forward it should ideally handle upgrade paths when permissions change and that be included as part of testing

**Comment by xmadscientist** [comment]
> Thanks!
> 
> My thinking so far is that I want to hear more from Raycast (e.g. Per) about how this is _supposed_ to work – if there could be some kind of interference between Slack integrations, and if so, whether that's considered a bug.
> 
> I also think I should add a text to the "Missing permissions" dialogue saying something to the effect of "If you keep seeing this message, please log out manually in the extension settings and uninstall all Raycast apps via Slack's admin UI".
> 
> I appreciate that this is a frustrating situation, but I'm still not clear on precisely what's going on, so I can't reproduce it myself to fix it more gracefully than that. But it does seem to me like there is some leakiness to (my assumptions about) the OAuth abstractions that extensions lean on.
> 
> It would be great if @thnk2wn and @xmadscientist could confirm whether doing those things fixes the issue. So specifically:
> 
> 1. In the "Slack Status" extension settings in Raycast, click "Logout" if available. (I think this might fix the "token_revoked" error.)
> 2. In the Slack web admin interface, select "Uninstall" on all "Raycast" apps.
> 3. Open "Slack Status" again and go through the auth flow.

This seemed to do it! Thanks so much!

**Comment by henrik** [comment]
@thnk2wn Thank you! How did you wipe the token manually?

I did try the upgrade path several times and spent a fair amount of time on it, but this seems to be some additional path of having multiple Raycast Slack apps that I still don’t understand well enough. Perhaps related to having both this extension and the other Slack extension?

@xmadscientist Thank you for confirming!

Happy holidays :)
